src/Interpreters/Set.cpp

```diff
@@ -254,8 +257,15 @@ ColumnPtr Set::execute(const Block & block, bool negative) const
 
     for (size_t i = 0; i < num_key_columns; ++i)
     {
-        checkTypesEqual(i, block.safeGetByPosition(i).type);
-        materialized_columns.emplace_back(block.safeGetByPosition(i).column->convertToFullColumnIfConst());
+        /// TODO: Optimize making cast only if types different
+        /// TODO: This case SELECT '1' IN (SELECT 1); should not work but with AccurateCastOrNull it works
```

Comment:
This is Ok :)

Reply:
BTW we also have obscure setting `transform_null_in`.

We can disable conversions (and resort to strict type checking) when it's enabled.

Reply:
Updated as we discussed, adding additional function accurateCast.